test/tests/dom/shadowDOMFunctions.js

```diff
@@ -0,0 +1,201 @@
+        it('should return shadow host if exists', () => {
+            const innerElement = document.querySelector('custom-web-component')?.shadowRoot?.querySelector('#inner-span');
+    
+            if (!innerElement) {
+                throw new Error('unable to find inner element');
+            }
+    
+            const result = getShadowHost(innerElement);
+            
+            if (!result) {
+                throw new Error(`should have returned the inner element, got undefined`);
+            }
+    
+            const hostId = result.getAttribute('id');
+    
+            if (hostId && hostId !== 'shadow-host') {
+                throw new Error(`should have returned 'shadow-host', got ${ hostId }`);
+            }
+    
+        });
+    });
+    
+    describe('insertShadowSlot cases', () => {
+        it('should throw exception if element is not in shadow DOM', () => {
+            const testElement = document.createElement('div');
+    
+            try {
+                insertShadowSlot(testElement);
+            } catch (error) {
+                if (!error.message.match(/Element is not in shadow dom/i)) {
+                    throw new Error(`should have thrown 'Element is not in shadow dom' exception, got: ${ error.message }`);
+                }
+            }
+    
+        });
+    
+        it('should throw exception if Host element is also in shadow dom', () => {
+    
+            // TestCase components setup
+            const customWrapper = document.createElement('custom-wrapper');
+            customWrapper.setAttribute('id', 'custom-wrapper');
+    
+            const customComponent = document.createElement('second-custom-web-component');
+            customComponent.setAttribute('id', 'shadow-host');
+    
+            // Get the div within custom-wrapper that will be used as inner host for a new shadow DOM
+            const hostDiv =  customWrapper.shadowRoot?.querySelector('#inner-host-div');
+    
+            const nestedShadow = hostDiv?.attachShadow({ mode: 'open' });
+            const innerSpan = document.createElement('span');
+            innerSpan.setAttribute('id', 'inner-span');
+            // $FlowFixMe
+            nestedShadow.appendChild(innerSpan);
+            // $FlowFixMe
+            customWrapper.appendChild(customComponent);
+            // $FlowFixMe
+            document.body.appendChild(customWrapper);
+
+            /**
+             * At this point the HTML structure looks like this:
+             * <html>
+             *    ...
+             *    <custom-wrapper>
+             *         #shadow-root (open)
+             *           <div id="wrapper-div">
+             *              #shadow-root (open)
+             *                <span id="inner-span"></span>
+             *           </div>
+             *    </custom-wrapper>
+             * </html>
+             */
+             
+    
+            try {
+                insertShadowSlot(innerSpan);
```

Comment:
I worry about false positives with this test. For example, if `insertShadowSlot()` does not throw an error, this test will still pass. What do you think about doing something like this so the test only passes when the error is thrown and passes the regex:

```js
let insertShadowSlotError;

try {
    insertShadowSlot(innerSpan);
} catch (error) {
    insertShadowSlotError = error;
}

if (!insertShadowSlotError?.error?.message.match(/Host element is also in shadow dom/)) {
    throw new Error(`should have thrown 'Host element is also in shadow dom' exception, got '${ error.message }'`);
}
```

Reply:
You´re  right, thanks for catching it, solved...

Reply:
Looks great! Can you use the same approach for the test above too? `should throw exception if element is not in shadow DOM`

Reply:
sure! fixed